### PR TITLE
[REVIEW] MD5 hashing list and null retention support

### PR DESCRIPTION
**Pull request by rwlee**
MD5 list column support, not adding support for nested lists at this point. Also augmenting existing functionality to optionally retain nulls rows in the resulting column if elements in that row are null.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/hash/hashing.cu [hunk 1]
```diff
@@ -679,7 +684,11 @@ std::unique_ptr<column> md5_hash(table_view const& input,
   auto chars_view = chars_column->mutable_view();
   auto d_chars    = chars_view.data<char>();
 
-  rmm::device_buffer null_mask{0, stream, mr};
+  rmm::device_buffer null_mask;
+  if (policy == retain_nulls::RETAIN_NULLS && input.num_columns() == 1)
+    null_mask = bitmask_and(input, mr, stream);
+  else
+    null_mask = rmm::device_buffer{0, stream, mr};
```
Comment (jrhemstad) [inline cpp/src/hash/hashing.cu:691]:
```suggestion
  rmm::device_buffer null_mask = [&](){
  if (policy == retain_nulls::RETAIN_NULLS && input.num_columns() == 1)
   return bitmask_and(input, mr, stream);
  else
    return rmm::device_buffer{0, stream, mr};
   }();
```

File: cpp/src/hash/hashing.cu [hunk 2]
```diff
@@ -679,7 +684,11 @@ std::unique_ptr<column> md5_hash(table_view const& input,
...
 
   bool const nullable     = has_nulls(input);
```
Comment (jrhemstad) [inline cpp/src/hash/hashing.cu:693]: nullable and has_nulls mean two different things.

File: cpp/src/hash/hashing.cu (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/hash/hashing.cu:691]: I don't understand this logic. It's calling `bitmask_and`, but only when there is a single column in the input?

File: cpp/include/cudf/hashing.hpp [hunk 3]
```diff
@@ -38,6 +43,7 @@ namespace cudf {
 std::unique_ptr<column> hash(
   table_view const& input,
   hash_id hash_function                     = hash_id::HASH_MURMUR3,
+  retain_nulls policy                       = retain_nulls::HASH_NULLS,
```
Comment (jrhemstad) [inline cpp/include/cudf/hashing.hpp:46]: I'm wondering if this API really needs this option. Options in APIs inherently complicate a user's comprehension and the implementation and therefore should only be added when strictly necessary. In this case, if `hash` always returns a non-nullable result (i.e., it hashes null values), then the `RETAIN_NULLS` policy can be accomplished trivially by the user by just calling `bitmask_and` on the input table and updating the null mask on the result.

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/hashing.hpp (hunk 3 above)
Comment (jrhemstad) [reply cpp/include/cudf/hashing.hpp:46]: Furthermore, adding this as an option requires that all code paths honor this option (i.e., the Murmur codepath would need to be updated too.

Comment (rwlee) [reply cpp/include/cudf/hashing.hpp:46]: I'll factor it out, makes some sense that the user modifies the resulting table. There's a hash value generated for the row either way, so modifying the column afterwards should be fine.

The most significant case is if we were to add functionality for an 'or' case -- where a null in any column of the row would create a null result. Skipping those rows could speed up the hash significantly, but that's not in the current scope of what we're trying to do.

File: cpp/src/hash/hashing.cu (hunk 1 above)
Comment (rwlee) [reply cpp/src/hash/hashing.cu:691]: Sorry, that was a fragment from when I was sketching out what alternate null policies would look like. Removing anyway

**Comment by harrism** [comment]
Since this is WIP, retargeting to 0.17. Please merge branch-0.17 into your PR branch @rwlee

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 4]
```diff
@@ -384,6 +384,49 @@ void testMD5HashMixed() {
...
+           "test MD5's message padding algorithm",
+           "hiJ\ud720\ud721\ud720\ud721", null);
+
+         ColumnVector result01 = ColumnVector.md5Hash(v0, v1);
+         ColumnVector result2 = ColumnVector.md5Hash(v2);
+         ColumnVector expected = ColumnVector.fromStrings(
+          "0cc175b9c0f1b6a831c399e269772661", "f5112705c2f6dc7d3fc6bd496df6c2e8",
+          "675c30ce6d1b27dcb5009b01be42e9bd", "8fa29148f63c1fe9248fdc4644e3a193",
+          "1bc221b25e6c4825929e884092f4044f", "d41d8cd98f00b204e9800998ecf8427e")) {
+      assertColumnsAreEqual(result01, expected);
```
Comment (razajafri) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:411]: expected and result01 should be flipped

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 5]
```diff
@@ -384,6 +384,49 @@ void testMD5HashMixed() {
...
+           "hiJ\ud720\ud721\ud720\ud721", null);
+         ColumnVector v2 = ColumnVector.fromStrings(
+           "a", "B\nc",  "dE\"\u0100\t\u0101 \u0500\u0501\\Fg2\'",
+           "A very long (greater than 128 bytes/char string) to test a multi hash-step data point " +
+           "in the MD5 hash function. This string needed to be longer.A 60 character string to " +
+           "test MD5's message padding algorithm",
+           "hiJ\ud720\ud721\ud720\ud721", null);
+
+         ColumnVector result01 = ColumnVector.md5Hash(v0, v1);
+         ColumnVector result2 = ColumnVector.md5Hash(v2);
```
Comment (razajafri) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:406]: result2 is not asserted

File: java/nvcomp_err.txt [hunk 6]
```diff
@@ -0,0 +1,318 @@
+[[1;34mINFO[m] Scanning for projects...
```
Comment (razajafri) [inline java/nvcomp_err.txt:1]: Is this supposed to be checked in?

File: java/rmm_log.txt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+[ 49623][22:21:21:223260][info  ] ----- RMM LOG BEGIN [PTDS DISABLED] -----
```
Comment (razajafri) [inline java/rmm_log.txt:1]: same comment, is this file needed to be checked in?

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 8]
```diff
@@ -970,6 +970,28 @@ public ColumnVector normalizeNANsAndZeros() {
...
 
+  /**
+   * Create a deep copy of the column while replacing the null mask. The resultant null mask is the
+   * bitwise merge of null masks in the columns given as arguments.
+   *
+   * @param mergeOp binary operator, currently only BITWISE_AND is supported.
+   * @param columns array of columns whose null masks are merged, must have identical number of rows.
+   * @return the new ColumnVector with merged null mask.
+   */
+  public ColumnVector mergeAndSetValidity(BinaryOp mergeOp, ColumnVector... columns) {
```
Comment (razajafri) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:981]: Do we really want it to be a list of columns that we want to merge the validity from? If its not a requirement then I would stick with just a single column to keep it simple. 

Also how about adding a method that takes in a hostbuffer to merge the validity from?

File: java/nvcomp_err.txt (hunk 6 above)
Comment (rwlee) [reply java/nvcomp_err.txt:1]: Nope, that's my bad.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 8 above)
Comment (rwlee) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:981]: It makes sense to merge a list of columns because the underlying cudf call wants a table, doesn't make sense to artificially limit that.

Makes sense to add a method that takes in a hostbuffer, but may be out of the scope and should be done as a separate follow up pr?

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java (hunk 5 above)
Comment (rwlee) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:406]: Missed this test, didn't modify it for list testing. Fixing now.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh [hunk 9]
```diff
@@ -49,89 +131,97 @@ void CUDA_DEVICE_CALLABLE uint32ToLowercaseHexString(uint32_t num, char* destina
...
+      if (!data_col.is_null(i)) {
+        T const& key = data_col.element<T>(i);
+        if (isnan(key)) {
+          T nan = std::numeric_limits<T>::quiet_NaN();
+          md5_process(nan, hash_state);
+        } else if (key == T{0.0}) {
+          md5_process(T{0.0}, hash_state);
+        } else {
+          md5_process(key, hash_state);
+        }
```
Comment (vuule) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:169]: Would be nice to have a helper like `normalize_nans_and_zeros_lambda` that we can call both here and in `MD5Hash::operator()`, instead of repeating the logic.

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java (hunk 8 above)
Comment (razajafri) [reply java/src/main/java/ai/rapids/cudf/ColumnVector.java:981]: Only if you feel there's an advantage to doing that. I vaguely remember we can do this in cudf.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh (hunk 9 above)
Comment (rwlee) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:169]: Would a straight helper function be appropriate so I can remove the repeating logic from murmur hash too?

**Review by jrhemstad — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/hash_functions.cuh (hunk 9 above)
Comment (rwlee) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:169]: Which one would you prefer?
```
template <typename T>
struct md5_normalize_nans_and_zeros_lambda {
  column_device_view data_col;
  md5_intermediate_data* hash_state;
  __device__ void operator()(int index)
  {
    T const& key = data_col.element<T>(index);
    if (isnan(key)) {
      T nan = std::numeric_limits<T>::quiet_NaN();
      md5_process(nan, hash_state);
    } else if (key == T{0.0}) {
      md5_process(T{0.0}, hash_state);
    } else {
      md5_process(key, hash_state);
    }
  }
};
```
vs.
```
template <typename TKey>
TKey CUDA_DEVICE_CALLABLE normalize_nans_and_zeros(TKey key) {
  if (isnan(key)) {
    return std::numeric_limits<TKey>::quiet_NaN();
  } else if (key == TKey{0.0}) {
    return TKey{0.0};
  } else {
    return key;
  }
}
```

Comment (vuule) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh:169]: I was thinking of the second option, but haven't looked into it too much. Whichever you think works best :)

Comment (rwlee) [reply cpp/include/cudf/detail/utilities/hash_functions.cuh]: Went with the second option, it is more reusable in future hash functions than the first. However, did not use it in the murmur hash normalization step, which uses a slightly different logic when it finds a `0.0` value.

File: cpp/include/cudf/detail/utilities/hash_functions.cuh [hunk 10]
```diff
@@ -106,6 +106,21 @@ void CUDA_DEVICE_CALLABLE md5_process(TKey const& key, md5_intermediate_data* ha
     hash_state->buffer_length = len - copylen;
   }
 }
+
+/**
+ * Normalization of floating point NANs and zeros helper
+ */
+template <typename TKey>
+TKey CUDA_DEVICE_CALLABLE normalize_nans_and_zeros_helper(TKey key)
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/utilities/hash_functions.cuh:114]: This should only be enabled for floating point types

**Comment by harrism** [comment]
@rwlee we can bring this forward to 0.16. You need to be careful though since this PR branch already has 0.17 commits in it. You will need to rebase and probably force push to get it back to 0.16 without those commits (merging branch-0.16 into your feature branch won't do it). First you need to click "edit" next to the title and change target to branch 0.16.

image

**Review by razajafri — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Comment by rwlee** [comment]
Rebased and forced pushed per @harrism's comment. Second force push necessary to drop an extra commit that included some 0.17 documentation changes.

**Review by harrism — APPROVED** [review]
CMake approval (there are no cmake changes)

**Review by karthikeyann — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 11]
```diff
@@ -980,18 +1002,27 @@ public static ColumnVector md5Hash(ColumnVector... columns) {
...
     for(int i = 0; i < columns.length; i++) {
       assert columns[i] != null : "Column vectors passed may not be null";
-      assert columns[i].getRowCount() == size : "Row count mismatch, all columns must have the same number of rows";
+      assert columns[i].getRowCount() == size : "Row count mismatch, all columns must be the same size";
       assert !columns[i].getType().isDurationType() : "Unsupported column type Duration";
       assert !columns[i].getType().isTimestamp() : "Unsupported column type Timestamp";
-      column_views[i] = columns[i].getNativeView();
+      if(columns[i].getType() == DType.LIST) {
+        assert columns[i].getNumChildren() == 1 : "Unrecognized list column format";
+        ColumnViewAccess listDataColumn = columns[i].getChildColumnViewAccess(0);
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:1015]: Can we make this a try with resources so we are guaranteed that it is closed?

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 12]
```diff
@@ -980,18 +1002,27 @@ public static ColumnVector md5Hash(ColumnVector... columns) {
...
+        assert !listDataColumn.getDataType().isDurationType() : "Unsupported column type List of Durations";
+        assert !listDataColumn.getDataType().isTimestamp() : "Unsupported column type List of Timestamp";
+        assert !listDataColumn.getDataType().isNestedType() : "Unsupported column type List of nested types";
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:1018]: nit: does the underlying hash code already test for these same assertions?  If so then we don't need to do it here.

**Review by revans2 — APPROVED** [review]
Just one nit really.
